test: add test cases for auction_test.go

jeonghoyeo7 opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

closes: #97 

## Tasks

- [x] Add various cases (only many, only worth, mix, some bidders, etc)
- [x] Add test cases for Refund
- [x] Add test cases for ReleaseSellingCoin
- [x] Suggest to modify `auction.go`
- [x] Modify `parseCoin` function to remove "_" in coin strings in `auction_test.go` and `bid_test.go`
- [x] Refactor the numbers in `parseCoin`  

## References

- 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Appropriate labels applied
- [ ] Targeted PR against correct branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://go.dev/blog/godoc).
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

jeonghoyeo7 committed:
feat: start to add test case for auction_test.go

jeonghoyeo7 committed:
test: add one case for batch auction

jeonghoyeo7 committed:
test: add one case for batch auction

jaybxyz committed:
fix: batch calculation logic

jeonghoyeo7 committed:
test: add test case for many

jeonghoyeo7 committed:
test: add test case for many

jeonghoyeo7 committed:
test: add test case for many

jeonghoyeo7 committed:
test: add more test cases for many/worth/mixed

jeonghoyeo7 committed:
test: add a test case for mixed bids without maxLimit

jeonghoyeo7 commented:
@kogisin  
I added one more case and will add more tonight.
- Question: `mInfo.MatchedLen` is the number of the matched bids, not the number of matched bidders, correct?
- Comment: Currently, in `placeBidBatchMany` and `placeBidBatchWorth`, maxBidAmt is added for the same bidder. So, it is not easy to setting maxBidAmt not to over the total Selling Amount. The current version is also okay to test, but we can think if there is a better way later.

jeonghoyeo7 commented on x/fundraising/keeper/auction.go:
```diff
@@ -490,14 +490,22 @@ func (k Keeper) CalculateBatchAllocation(ctx sdk.Context, auction types.AuctionI
 	// Iterate from the highest matching bid price and stop until it finds
 	// the matching information to store them into MatchingInfo object
-	for _, bid := range bids {
+	for _, bid := range bids { // (1, 500), (0.9, 500), (0.8, 500)
```
I will remove the comment later.

jaybxyz replied:
Yes, when this PR gets R4R I suggest you to remove all the unnecessary comments and unused codes.

jeonghoyeo7 commented:
@kogisin One more thing is about maxBidAmount. Currently, FixedPriceAuction does not allowe bidders to bid with more than maxBidAmount. But BatchAuction allows that a bidder bids more than maxBidAmount. It can be understood in that a batch auction calculates allocations when finished. However, both Many and Worth type bids make a bidder get the selling coing as at least it bids, if the bid wins. In this regards, how about restrict a bid not to have more than maxBidAmount?
For example, a Many type bid cannot bid with more than maxBidAmount. A Worth type bid cannot bid with the price and payingcoin that the payingCoing divided with the bid price is not over maxBidAmount.

jaybxyz commented:
I will reply to your question and comments offline.

jeonghoyeo7 committed:
test: add a test case for mixed bids with maxBidAmount

jeonghoyeo7 commented:
@kogisin In the last test case I added, I found the case that, even though a bid has the bid price higher than the matched price, the bidder does not get any Selling coin from the bid due to the limitation from `maxBidAmount`. This means, `MatchedLen` just means the number of bid having higher than or equal to the matched price. `MatchedLen` does not mean either the number of WinningBidders or the number of winning bids that offers a non-zero selling coin. I don't think this should be fixed, but we can just be aware of this kind of cases and definition of `MatchedLen`.

jeonghoyeo7 committed:
feat/test: modify test cases and suggest for auction.go

jeonghoyeo7 committed:
feat/test: modify test cases and suggest for auction.go

jeonghoyeo7 commented:
@kogisin During modifying test cases, I found some point that might be needed to modify for `auction.go`. It seems non-zero allocation is also added as {input, output}
```
	for bidder, allocAmt := range mInfo.AllocationMap {
		if allocAmt.IsZero() {
			continue
		}
		allocateCoins := sdk.NewCoins(sdk.NewCoin(sellingCoinDenom, allocAmt))
		bidderAddr, _ := sdk.AccAddressFromBech32(bidder)

		inputs = append(inputs, banktypes.NewInput(sellingReserveAddr, allocateCoins))
		outputs = append(outputs, banktypes.NewOutput(bidderAddr, allocateCoins))
	}
```
I added `	if allocAmt.IsZero() {continue}`
Please check this out and see if this is correct, and please feel free to modify and fix.

jeonghoyeo7 commented:
@kogisin `ReleaseRemainingSellingCoin`, `RefundPayingCoin`, and `AllocateVestingPayingCoin` might have the same issue for zero allocation.

jeonghoyeo7 committed:
feat/test: modify the first test case and suggest for auction.go

jeonghoyeo7 commented on x/fundraising/keeper/auction.go:
```diff
@@ -42,6 +42,10 @@ func (k Keeper) RefundPayingCoin(ctx sdk.Context, auction types.AuctionI, mInfo
 
 	// Refund the unmatched bid amount back to the bidder
 	for bidder, refundAmt := range mInfo.RefundMap {
+		if refundAmt.IsZero() {
```
@kogisin Please check if this is a proper addition.

jaybxyz replied:
Yes, I think this is a proper addition since zero refund amount can exist for the bids that are fully matched.

jeonghoyeo7 commented on x/fundraising/keeper/auction.go:
```diff
@@ -71,6 +75,9 @@ func (k Keeper) AllocateSellingCoin(ctx sdk.Context, auction types.AuctionI, mIn
 	// Allocate coins to all matched bidders in AllocationMap and
 	// set the amounts in trasnaction inputs and outputs from the selling reserve account
 	for bidder, allocAmt := range mInfo.AllocationMap {
+		if allocAmt.IsZero() {
```
@kogisin Please check if this is a proper addition.

jaybxyz replied:
Yes, I think this is a proper addition.

jeonghoyeo7 committed:
test: modify the test cases to check allocation function

jeonghoyeo7 committed:
Merge branch 'main' into 97-add-test-cases-for-auctions

* main:
  test: add test cases for maxBidAmount in BatchAuction
  feat: add validate check for batch auction with maxBidAmount

jeonghoyeo7 committed:
test: modify the test cases according to maxBidAmount

jeonghoyeo7 committed:
test: refactor the numbers with parseCoin and remove unnecessary test in auction_test.go
